### ISSUE TITLE
Roadmap to improve the documentation of tactics.

### ISSUE THREAD

**Issue by Zimmi48** [coq/coq#8946]
Here are some of the things that need to be done IMO:

- [ ] Change the general structure of the Tactics chapter:
  - [ ] Reorganize the sections "Applying theorems" and "Controlling the proof flow" in a part on "Concluding tactics" (`assumption`, `exact`, `reflexivity`, `discriminate`, `contradiction` and `easy`), a part on "Backward reasoning" (with `apply`, `enough`, `f_equal`?) and a part on "Forward reasoning" (consolidating the doc of `apply in`, `specialize`, `pose proof`, `assert`; we critically lack an explicit part on this alternative kind of proof organization that is perfectly supported by Coq but rarely used).
  - [ ] Consolidate the "Rewriting" and the "Performing computations" section in a single section on applying term transformations.
  - [ ] Rename the section "Classical tactics" in "Tactics for classical logic" or "Tactics requiring additional axioms" (and then consolidate with other such tactics).
  - [ ] Move the section "Checking properties of terms" in the Ltac chapter.
  - [ ] Move the section "Non-logical tactics" to the Proof handling chapter.
  - [ ] Remove the "Simple macro tactics" section.
  - [ ] Consolidate the "Decision procedure", "Automating" sections and the Omega, Micromega, ring and field, and nsatz chapters into a single "Decision procedures" chapter.
  - [ ] Consolidate the "Automation" and "Controlling automation" sections and the relevant parts of the Typeclass chapter into a single "User-controlled automation" chapter.
  - [ ] Merge the "Detailed examples of tactics" chapter in the Tactics chapter.
- [ ] Deprecate tactics that are not used anymore: `cut`, `elim`, `case` (vanilla versions), as well as `simple destruct`.
- [ ] Identify and deprecate useless tactics (example: `lapply`).
- [ ] Add as many cross-links as possible between various sections of the Tactics chapter and between the Tactics chapter and the SSReflect chapter.
- [ ] Add many examples.
- [ ] Add warnings about known inconsistencies and known problems so that users are not left wondering.

I am opening this issue after a discussion with @jfehrle. There was also @bollu (IIRC) who had asked me about writing some of the needs regarding documentation.
Feel free to contribute to part of this. Please state it in the issue thread so that we are aware you are working on it. Also feel free to mention additional changes that you feel would be necessary (regarding documentation). I'll open more similar documentation roadmap issues in the future and will add them all to the Documentation project.

**Comment by jfehrle** [comment]
> Please state it in the issue thread

Meaning this thread?

**Comment by JasonGross** [comment]
> Deprecate tactics that are not used anymore: cut, elim, case (vanilla
versions),

>
What are suggested replacements for these?  I never use elim, but I often
use cut rather than assert; revgoals(; revert), and case is essential for
the rare cases where my hypotheses are dependent enough that destruct
fails, but my goal is simple enough that case does not.

**Comment by Zimmi48** [comment]
The replacement for `assert; revgoals` is `enough`. Strangely, many people seems not to know about this tactic.
I was not aware of a difference of power between `case` and `destruct`. Could you provide an example?

**Comment by jfehrle** [comment]
>  Keeping things in alphabetical index is important for easy retrieval.

Huh?  Adding a suffix like "(tactical)", "(Ltac)", etc. won't change the order in the index.  (Though eventually we may want to add brief overview summaries of tactics to the beginning of sections so you don't need to read/skim the entire section as often.)

I'll look at doing some of the easier items you mention in https://github.com/coq/coq/issues/8946, maybe as a separate commit/PR.

**Comment by herbelin** [comment]
Most of everything above looks very good and I wish that everyone can contribute what they think is important to say and feel represented in the final result.

In case working at the documentation shows blatant inconsistencies of behavior here and there, and I know there are, I can certainly help on the implementation side. I have (quite many) open branches destined to extend or harmonize the behavior of tactics in general and I'd be happy to collaborate at improving things in parallel to their documentation.

**Comment by jfehrle** [comment]
I just generated a PDF of the tactics file, which comes to 118 pages just to get a printout of the 6 or 7 pages I'm revising.  We should consider breaking up the chapters into more, smaller HTML files--maybe up to 10 pages each because a) the beginning and end of section is not very apparent b) better for the browser search function c) the current files take a noticeable time to load and d) less merging as people update sections.

**Comment by anton-trunov** [comment]
> Deprecate tactics that are not used anymore: cut, elim, case (vanilla versions), as well as simple destruct.

@Zimmi48 Perhaps we should remove those tactics from the standard library first? I expect this to require quite a bit of work.

**Comment by Zimmi48** [comment]
Absolutely. This could be of noticing if there are any troubles of the kind mentioned by Jason.

**Comment by cpitclaudel** [comment]
> Deprecate tactics that are not used anymore: cut, elim, case (vanilla versions), as well as simple destruct.

I'd be careful with these: in some cases, ``case`` is twice as fast as ``destruct``.

**Comment by cpitclaudel** [comment]
Here's an additional idea that's been on my plate for a while:

Tactics definitions could be annotated with various flags, indicating what kind of behavior they have.  For example, tactics like `exact`, `assumption`, and `omega` either solve the goal or fail (PG highlights them in red); conversely, `auto`, `trivial`, `intuition` attempt to solve the goal, but always succeed. Lastly, tactics like `simpl`, `cbv`, or `unfold` always succeed, even if they don't make progress. And `admit` is in a class of its own, of course.

Multiple syntax highlighters for Coq already have these or similar categories, but each highlighter has a slightly different and outdated view of the tactics.  It would be good to have one canonical source for this information (FWIW, PG currently distinguishes between "solve" tactics, "cheat" tactics, regular tactics, and tacticals; Pygments distinguishes between "terminators", "tactics", and "control").

So, the concrete proposal is to add a new option to `.. tacn`, called `:tags:`, which for now could take only the following values: `solve`, `cheat`.

Another thing I'd like to do is to introduce a new subclass of `tacn::` called `tactical`, for things like `try`, `repeat`, `first`, etc.

**Comment by Zimmi48** [comment]
These are good ideas. About the last one, have you noted #9179? It has been sort of blocked on finding the right definition for a tactical, I guess. Mine is quite restrictive and includes the examples you just gave but Jim was including much more.

Related to this is the question of what to put in the main Ltac grammar. Currently, it contains most (all?) tacticals. I don't find this particularly useful, especially if they are then repeated as `.. tacn::`. The Ltac2 documentation follows another approach (https://coq.github.io/doc/master/refman/proof-engine/ltac2.html) but it might just be because its documentation is incomplete ATM.

**Comment by jfehrle** [comment]
> [#9179] has been sort of blocked on finding the right definition for a tactical, I guess.

Not entirely sure what I'll do with that PR.  I transitioned to a related problem, extracting the grammar in order to document it (#9884, which I want to get back to soon).  FWIW, there are other things worth indexing, such as examples and concepts.

**Comment by cpitclaudel** [comment]
> Related to this is the question of what to put in the main Ltac grammar. Currently, it contains most (all?) tacticals. I don't find this particularly useful, especially if they are then repeated as .. tacn::. The Ltac2 documentation follows another approach (https://coq.github.io/doc/master/refman/proof-engine/ltac2.html) but it might just be because its documentation is incomplete ATM.

I don't find the full grammar very useful either, but it's not necessarily bad to have it for reference.
I'm not sure I understand what approach the Ltac2 chapter takes (plain code blocks with 'val'?)

Another style is the one used in the ssreflect chapter (`.. prodn::` for most things).  One problem with that is that it doesn't distinguish neatly whether the syntax is gallina, Ltac, or something else (you have to follow multiple levels of indirection: for example `r_pattern ::= @term %| @c_pattern %| in {? @ident in } @term` is tactic syntax, because `r_pattern` is used is `r_prefix`, which is used in `rstep`, which is used in the `rewrite` tactic).

> EDIT: more about Ltac2: Pierre-Marie argues that many non-primitive things shouldn't get documented in the refman but in the documentation generated with coqdoc from the .v files instead. I would personally like if the two could be closer anyway. Any thoughts?

I agree.  The way people do this with sphinx is "autodoc", where Sphinx goes and parses relevant docstrings out of `.py` files.  If we can agree on a proper in-code documentation format, we should do the same (it was one of the nice things we could do if we had Sphinx that we originally discussed).  The implementation could of course be written in OCaml, dumping structured data to be read by Sphinx, if that's preferable.

**Comment by cpitclaudel** [comment]
One more thought: there are multiple constructs that are not documented on their own: for example, the `let cofix` form is only documented in a ``.. productionlist:: coq``.

Another problem with these ``.. productionlist::`` constructs is that they don't actually say which language they are for; for example the ltac one is labeled ``.. productionlist::  coq``.

**Comment by jfehrle** [comment]
> I don't find the full grammar very useful either, but it's not necessarily bad to have it for reference.

I think we must include the full grammar in _some_ form for completeness.  And we need to figure out how best to present it.

> Another problem with these `.. productionlist::` constructs is that they don't actually say which language they are for; for example the ltac one is labeled `.. productionlist:: coq`.

My recollection is that if you have 2 productionlists with different tags, such as "coq" and "foo", you don't hyperlinks that operate across them.

> (@Zimmi48) Pierre-Marie argues that many non-primitive things shouldn't get documented in the refman but in the documentation generated with coqdoc from the .v files instead.

What are these non-primitive things?  If they are part of the grammar, they should be in the reference manual.

> Another style is the one used in the ssreflect chapter (`.. prodn::` for most things).

I don't see how you'd use `prodn` for nonterminals that have multiple productions.  Also, it seems undesirable to use different names for nonterminals in the doc compared to the grammar.  The ssr nonterminals mostly (entirely?) start with "ssr" in the grammar but not in the doc.

Furthermore, the doc grammar doesn't match the mlg file, for example:

![image](https://user-images.githubusercontent.com/1253341/57589086-9c41f300-74d3-11e9-9cd9-609c0090cb8e.png)

https://coq.inria.fr/distrib/current/refman/proof-engine/ssreflect-proof-language.html#coq:tacn.rewrite-ssreflect

doesn't match what's in `ssrparser.mlg`:

```
TACTIC EXTEND ssrrewrite
  | [ "rewrite" ssrrwargs(args) ssrclauses(clauses) ] ->
    { tclCLAUSES (old_tac (ssrrewritetac ist args)) clauses }
END
```

and there is no `rstep` in the .mlg!

I think there's a lot of detail work needed to accurately describe the grammar in the doc.  @cpitclaudel, please take a look at #9884, let me know if you think that's a promising approach  (It will need a good bit of refinement.)

FWIW, we seem to be getting into a deep dive on this topic in a very broad issue.

**Comment by cpitclaudel** [comment]
> I don't see how you'd use prodn for nonterminals that have multiple productions.

I think `prodn` allows `nont += …`

> My recollection is that if you have 2 productionlists with different tags, such as "coq" and "foo", you don't get hyperlinks that operate across them.

I think that's correct.  I'm not a huge fan of `.. productionlist`, all in all (its syntax is ambiguous, too, which doesn't help :'(

> Also, it seems undesirable to use different names for nonterminals in the doc compared to the grammar. The ssr nonterminals mostly (entirely?) start with "ssr" in the grammar but not in the doc.

Agreed, but that's orthogonal to using `.. prodn` vs `.. productionlist` vs …, isnt' it?

> I think there's a lot of detail work needed to accurately describe the grammar in the doc. @cpitclaudel, please take a look at #9884, let me know if you think that's a promising approach (It will need a good bit of refinement.)

Will do.  As a meta-level point, though, I think discrepancies between the documented grammar and the actual grammar are not necessarily a problem, since often there can be deprecated syntax, compatibility syntax, or technical kludges plaguing the grammar that don't need to be reflected in the manual.

**Comment by jfehrle** [comment]
> there can be deprecated syntax, compatibility syntax, or technical kludges plaguing the grammar that don't need to be reflected in the manual.

Or those could be marked as such.

**Comment by Zimmi48** [comment]
>  If they are part of the grammar, they should be in the reference manual.

Some of them are defined using notations (user-level). Do you call this “part of the grammar” or not? But some of them are already in the manual actually.

Some others are just definitions, so definitely not part of the grammar.

> The ssr nonterminals mostly (entirely?) start with "ssr" in the grammar but not in the doc.

They probably should.

**Comment by jfehrle** [comment]
> Some of them are defined using notations (user-level). Do you call this “part of the grammar” or not? But some of them are already in the manual actually.

The doc should either include them or link to a easily-searched document describing them, for example at the beginning of the tactics section.  Or maybe include a subset in the doc, perhaps identifying them as notations.  "Tactic Notation" appears in the .v files 148 times.  ("Notation" appears 2250 times.  We should look at tactics first but also consider notations in general later.) 

There should be an easy, documented way to find a tactic notation definition given its name.  There should also be a reasonable way to discover tactic notations that may be relevant to the user.  These procedures should be explained in the doc.  Also, the tactic notations should be organized by topic rather than just alphabetically.  And identify which notations are important for users.

My initial impression of coqdoc is that it's not very user-friendly or useful.  For example, I don't see a list of tactics defined as notations.  The indexing is also strange.  For example, clicking on "Notation Index   A" in https://coq.inria.fr/distrib/current/stdlib/genindex.html doesn't show you the A entries, and it seems the same for the other letters.  Seems to me we should plan to significantly improve coqdoc.  A simpler step we could take in the short term is to generate an index of tactic notations.  Plus adding doc comments in the .v files that can appear in coqdoc.  (Is there an easy way to find the definition of, say `+!`?)

```
A (notation)
_ * _ [in Coq.Numbers.NatInt.NZAxioms]
_ - _ [in Coq.Numbers.NatInt.NZAxioms]
_ + _ [in Coq.Numbers.NatInt.NZAxioms]
[ _ ] [in Coq.setoid_ring.Ring_theory]
?=! [in Coq.setoid_ring.Ring_theory]
-! _ [in Coq.setoid_ring.Ring_theory]
_ -! _ [in Coq.setoid_ring.Ring_theory]
_ *! _ [in Coq.setoid_ring.Ring_theory]
_ +! _ [in Coq.setoid_ring.Ring_theory]
- _ [in Coq.setoid_ring.Ring_theory]
_ - _ [in Coq.setoid_ring.Ring_theory]
_ - _ [in Coq.setoid_ring.Ring_theory]
- _ [in Coq.setoid_ring.Ring_theory]
_ * _ [in Coq.setoid_ring.Ring_theory]
_ + _ [in Coq.setoid_ring.Ring_theory]
_ == _ [in Coq.setoid_ring.Ring_theory]
1 [in Coq.setoid_ring.Ring_theory]
0 [in Coq.setoid_ring.Ring_theory]
```

Of course, FWIW, doc_grammar doesn't have the notation information.

**Comment by jfehrle** [comment]
There are also 724 occurrences of "Ltac".  The considerations are similar to tactic notations, though there's no index for Ltac definitions in coqdoc.  Maybe we could add a tactic index containing both.

**Comment by Zimmi48** [comment]
Indeed, the issue is that coqdoc has been mostly untouched for a very long time, so it hasn't really evolved with the needs. If you're willing to spend time improving it, that would be definitely welcome. In case of doubt about the design, do not hesitate to open an issue to discuss it first.

**Comment by jfehrle** [comment]
Perhaps later on.  There's plenty to do on the main doc first.

**Comment by jfehrle** [comment]
> > Deprecate tactics that are not used anymore: cut, elim, case (vanilla versions), as well as simple destruct.
> 
> @cpitclaudel I'd be careful with these: in some cases, `case` is twice as fast as `destruct`.

Is that a significant impact on users?  (e.g. If one takes .05 seconds and the other .1 seconds, it will be imperceptible for interactive use.  How about when you run a script?)

**Comment by Zimmi48** [comment]
For most users, this is clearly insignificant, but some users like @cpitclaudel or @JasonGross tend to use a heavily automated proof style (see http://adam.chlipala.net/cpdt/) with lots of proof attempts and where performance of these basic tactics matter a lot.
